internal/pkg/blockdevice/probe/probe.go

```diff
@@ -52,18 +52,23 @@ func All() (probed []*ProbedBlockDevice, err error) {
 			// file system exists without a partition table.
 			if sb = probe(devpath); sb != nil {
 				probed = append(probed, &ProbedBlockDevice{BlockDevice: bd, SuperBlock: sb, Path: devpath})
+			} else {
+				// No need to close the block device here since the underlying file is closed on an error.
```

Comment:
I don't think I'm able to follow all the logic in this function at the momemnt, but does it make sense to split it up so that we can do more simple `defer bd.Close()` to ensure it gets closed?

just my 2 cents here, I might be totally wrong as well

Reply:
The goal of this function is to return a slice of `BlockDevice`s. The `BlockDevice` type has an underlying `os.File` that we nee to keep open. This function probes all block devices for a valid file system, and if a block device doesn't have a valid file system, we should close the underlying file by calling `bd.Close()`, but we don't to close it if the block device has a valid file system.

Reply:
got it, thanks

I wonder if we could split the probe and what we return so that we open & always close for the probe? idk, just a thought to make it less prone to bugs.

I don't mean to block this PR, but just as an idea

Reply:
I agree, it needs some rethinking. Anything in the standard library you can think of that we can get inspiration from?

Reply:
very sketchy (without error handling):

(one extra open for valid fs block devices)

```


for _, devName := range deviceNames {
    if probeOk(devName) {
       devFd := open(devName)
       devices = append(devices, devFd)
    }

...

function probeOk(devName string) bool {
   devFd := open(devName)
   defer devFd.Close()
}
```